source/RoslynAnalyzers/ApiControllerAnalyzer.cs

```diff
@@ -0,0 +1,160 @@
+﻿using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Linq;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+
+namespace Octopus.RoslynAnalyzers;
+using static Descriptors;
+
+[DiagnosticAnalyzer(LanguageNames.CSharp)]
+public partial class ApiControllerAnalyzer : DiagnosticAnalyzer
+{
+    public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(
+        MustNotHaveSwaggerOperationAttribute,
+        MustNotReturnActionResults);
+
+    public override void Initialize(AnalysisContext context)
+    {
+        context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+        if(!Debugger.IsAttached) context.EnableConcurrentExecution();
+        context.RegisterSyntaxNodeAction(CheckNode, SyntaxKind.ClassDeclaration);
+    }
+
+    void CheckNode(SyntaxNodeAnalysisContext context)
+    {
+        if (context.Node is not ClassDeclarationSyntax classDec || !IsApiController(context, classDec)) return;
+  
+        var attrNames = classDec.AttributeLists.SelectMany(al => al.Attributes.Select(a => (a.Name as IdentifierNameSyntax)?.Identifier.Text));
+        var isExperimental = attrNames.Any(a => a == "Experimental");
+
+        var actionMethods = new List<MethodDeclarationSyntax>();
+        foreach (var methodDec in classDec.DescendantNodes().OfType<MethodDeclarationSyntax>())
+        {
+            // only validate action methods
+            if(!IsActionMethod(context, methodDec)) continue;
+            actionMethods.Add(methodDec);
+
+            if (!isExperimental)
+            {
+                MustNot_HaveSwaggerOperationAttribute(context, methodDec);
+            }
+
+            MustNot_ReturnActionResults(context, methodDec);
+        }
+
+        // ControllersAcceptingNewPayloads_MustBeNamedCorrectly();
+    }
+
+    static bool MustNot_HaveSwaggerOperationAttribute(SyntaxNodeAnalysisContext context, MethodDeclarationSyntax methodDec)
+    {
+        var swaggerOperationAttr = methodDec.AttributeLists.SelectMany(al => al.Attributes).FirstOrDefault(al => al.Name is IdentifierNameSyntax { Identifier.Text: "SwaggerOperation" });
```

Comment:
Tiny perf nitpick: we materialised and allocated this collection in the parent, we could pass the same collection down here to inspect for `SwaggerOperation` instead of allocating again.

Reply:
This one is interesting.
The compiler+JIT can inline the SelectMany+FirstOrDefault and turn this into a simple non-allocating loop over structures that already exist.

If they do, then it's better to leave the code as is.
If they don't, then it's better to allocate a temporary array rather than calling SelectMany twice,

What does it actually do? It's probably not smart enough to inline them, but proof would require benchmarks etc. Given the difference at this point is very minor I'd like to leave it as-is and maybe take a run at it later when adding more analyzers?

Reply:
Sounds fine :)